### PR TITLE
Multiwindow Desktop Streamer

### PR DESCRIPTION
**Pull request by eile**

### CODE REVIEW

**Comment by eile** [comment]
@rdumusc FYI

**Comment by eile** [comment]
Missing is the window raise, but I leave that for later.

File: apps/DesktopStreamer/AppNapSuspender.h [hunk 1]
```diff
@@ -34,7 +34,7 @@
 /* The views and conclusions contained in the software and           */
 /* documentation are those of the authors and should not be          */
 /* interpreted as representing official policies, either expressed   */
-/* or implied, of The University of Texas at Austin.                 */
+/* or implied, of The Ecole Polytechnique Federal de Lausanne.       */
```
Comment (rdumusc) [inline apps/DesktopStreamer/AppNapSuspender.h]: I'm not sure if we should use the accents, but we definitely should put a final e: FédéralE

**Comment by eile** [comment]
Hold merge.

**Comment by eile** [comment]
good to go

File: apps/DesktopStreamer/MainWindow.cpp [hunk 2]
```diff
@@ -177,205 +149,189 @@ void MainWindow::closeEvent( QCloseEvent* closeEvt )
...
+#ifdef DEFLECT_USE_QT5MACEXTRAS
+    const QModelIndexList windowIndices =
+        _listView->selectionModel()->selectedIndexes();
 
-    _stream.reset(
-        new deflect::Stream( _streamnameLineEdit->text().toStdString(),
-                             streamHost.toStdString( )));
-    if( !_stream->isConnected( ))
+    StreamMap streams;
+    for( QPersistentModelIndex index : windowIndices )
```
Comment (dnachbaur) [inline apps/DesktopStreamer/MainWindow.cpp]: const&

File: apps/DesktopStreamer/MainWindow.cpp [hunk 3]
```diff
@@ -177,205 +149,189 @@ void MainWindow::closeEvent( QCloseEvent* closeEvt )
...
+            _listView->model()->data( index, Qt::DisplayRole ).
+                toString().toStdString() : std::string();
+        const std::string streamName = std::to_string( ++_streamID ) +
+                                       " " + name + " - " +
+                                      _streamnameLineEdit->text().toStdString();
+        StreamPtr stream( new Stream( *this, index, streamName, host ));
+
+        if( !stream->isConnected( ))
+        {
+            _statusbar->showMessage( std::string( "Could not connect to host " +
```
Comment (dnachbaur) [inline apps/DesktopStreamer/MainWindow.cpp]: QString

File: apps/DesktopStreamer/MainWindow.cpp [hunk 4]
```diff
@@ -177,205 +149,189 @@ void MainWindow::closeEvent( QCloseEvent* closeEvt )
...
         return;
     }
-#endif
 
-#ifdef STREAM_EVENTS_SUPPORTED
-    if( _desktopInteractionCheckBox->isChecked( ))
-        _stream->registerForEvents();
+    if( _streams.empty( ))
+    {
+        const QPersistentModelIndex index;
```
Comment (dnachbaur) [inline apps/DesktopStreamer/MainWindow.cpp]: what does this uninitialized index do?

File: apps/DesktopStreamer/MainWindow.cpp [hunk 5]
```diff
@@ -177,205 +149,189 @@ void MainWindow::closeEvent( QCloseEvent* closeEvt )
...
+#ifdef DEFLECT_USE_QT5MACEXTRAS
+void MainWindow::_deselect( ConstStreamPtr stream )
 {
-    // frame rate limiting
-    const int maxFrameRate = _maxFrameRateSpinBox->value();
-    const int desiredFrameTime = (int)( 1000.f * 1.f / (float)maxFrameRate );
-    const int sleepTime = desiredFrameTime - elapsedFrameTime;
-
-    if( sleepTime > 0 )
+    QPersistentModelIndex index = stream->getIndex();
```
Comment (dnachbaur) [inline apps/DesktopStreamer/MainWindow.cpp]: const&

File: apps/DesktopStreamer/MainWindow.h [hunk 6]
```diff
@@ -62,28 +65,26 @@ class MainWindow : public QMainWindow, public Ui::MainWindow
     MainWindow();
     ~MainWindow();
 
+    const QAbstractItemModel* getWindowModel() const { return _listView->model(); }
```
Comment (dnachbaur) [inline apps/DesktopStreamer/MainWindow.h]: getItemModel()

File: apps/DesktopStreamer/Stream.cpp [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+{}
+
+Stream::~Stream()
+{}
+
+bool Stream::processEvents( const bool interact )
+{
+    while( isRegisteredForEvents() && hasEvent( ))
+    {
+        deflect::Event event = getEvent();
```
Comment (dnachbaur) [inline apps/DesktopStreamer/Stream.cpp:77]: where did the retina mouse coordinate compensation go?

File: apps/DesktopStreamer/Stream.h [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+    /**
+     * Process all pending events.
+     *
+     * @param interact enable interaction from the server
+     * @return true on success, false if the stream should be closed.
+     */
+    bool processEvents( bool interact );
+
+    QPersistentModelIndex getIndex() const { return _window; }
```
Comment (dnachbaur) [inline apps/DesktopStreamer/Stream.h]: const&

File: doc/Changelog.md [hunk 9]
```diff
@@ -1,7 +1,9 @@
 Changelog {#Changelog}
 ============
 
-## Deflect 0.9 (git master)
+## Deflect 0.10 (git master)
+* [79](https://github.com/BlueBrain/Deflect/pull/79):
+  DesktopStreamer: Support multi-window streaming on OS X
```
Comment (dnachbaur) [inline doc/Changelog.md]: ..fix retina event handling, fix crash on stream close from DisplayCluster

File: apps/DesktopStreamer/MainWindow.cpp (hunk 3 above)
Comment (eile) [inline apps/DesktopStreamer/MainWindow.cpp]: QString can't '+ host'

File: apps/DesktopStreamer/MainWindow.cpp (hunk 4 above)
Comment (eile) [inline apps/DesktopStreamer/MainWindow.cpp]: invalid == Desktop == Index( row( 0 )). Same magic as before. All indices with row < 0 are considered desktop in the stream.

File: apps/DesktopStreamer/Stream.cpp (hunk 7 above)
Comment (eile) [inline apps/DesktopStreamer/Stream.cpp:77]:
```
QRect _windowRect; // position on host in non-retina coordinates

// render mouse cursor
const QPoint mousePos = ( QCursor::pos() - _windowRect.topLeft( )) *
                        _parent.devicePixelRatio() -
                        QPoint( _cursor.width()/2, _cursor.height()/2 );
```

**Comment by eile** [comment]
All others done.

File: apps/DesktopStreamer/MainWindow.cpp (hunk 3 above)
Comment (dnachbaur) [inline apps/DesktopStreamer/MainWindow.cpp]: QString can do more than std::string => QString("%1").arg(host)

File: apps/DesktopStreamer/MainWindow.cpp (hunk 4 above)
Comment (dnachbaur) [inline apps/DesktopStreamer/MainWindow.cpp]: ugh. Comment then :)

File: apps/DesktopStreamer/Stream.cpp (hunk 7 above)
Comment (dnachbaur) [inline apps/DesktopStreamer/Stream.cpp:77]: That's rendering, not event handling imo

File: apps/DesktopStreamer/MainWindow.cpp (hunk 3 above)
Comment (eile) [inline apps/DesktopStreamer/MainWindow.cpp]: "more" ;)

File: apps/DesktopStreamer/Stream.cpp (hunk 7 above)
Comment (eile) [inline apps/DesktopStreamer/Stream.cpp:77]: right. Events are always processed in non-retina space := no conversion needed. Only the cursor rendering on the retina pixmap needs this. KISS

File: apps/DesktopStreamer/MainWindow.cpp (hunk 3 above)
Comment (eile) [inline apps/DesktopStreamer/MainWindow.cpp]: Does not even take a std::string as arg :100:

Comment (dnachbaur) [inline apps/DesktopStreamer/MainWindow.cpp]: QString::fromStdString()

**Pull request merged**